### PR TITLE
SW-3809 - Backend- Select materials csv based on laser cutter mode

### PR DESCRIPTION
**Pull request by irlaec**

### CODE REVIEW

File: octoprint_mrbeam/util/material_csv_parser.py [hunk 1]
```diff
@@ -16,6 +16,15 @@
...
 VALID_MODELS = [MRBEAM, MRB_DREAMCUT, MRB_DREAMCUT_S, MRB_READY, MRB_DREAMCUT_X]
 
+DEFAULT_MODE = "default"
+ROTARY_MODE = "rotary"
+
+MATERIALS_CSV_DIR = "files/material_settings/"
+MATERIALS_CSVS = {
+    DEFAULT_MODE: MATERIALS_CSV_DIR + "materials.csv",
+    ROTARY_MODE: MATERIALS_CSV_DIR + "materials_rotary.csv",
+}
```
Comment (khaledsherkawi) [inline octoprint_mrbeam/util/material_csv_parser.py]:
```suggestion
MATERIALS_CSV_DIR = "files/material_settings/"
DEFAULT_MATERIALS_CSV_DIR = MATERIALS_CSV_DIR + "materials.csv"
ROTARY_MATERIALS_CSV_DIR = MATERIALS_CSV_DIR + "materials_rotary.csv"
MATERIALS_CSVS = {
    DEFAULT_MODE: DEFAULT_MATERIALS_CSV_DIR,
    ROTARY_MODE: ROTARY_MATERIALS_CSV_DIR,
}
```

File: octoprint_mrbeam/util/material_csv_parser.py [hunk 2]
```diff
@@ -16,6 +16,15 @@
 
 VALID_MODELS = [MRBEAM, MRB_DREAMCUT, MRB_DREAMCUT_S, MRB_READY, MRB_DREAMCUT_X]
 
+DEFAULT_MODE = "default"
+ROTARY_MODE = "rotary"
```
Comment (khaledsherkawi) [inline octoprint_mrbeam/util/material_csv_parser.py:21]:
```suggestion
# TODO: import these from mode services
DEFAULT_MODE = "default"
ROTARY_MODE = "rotary"
```

File: octoprint_mrbeam/materials.py [hunk 3]
```diff
@@ -1,12 +1,16 @@
...
-from octoprint_mrbeam.mrb_logger import mrb_logger
-from octoprint_mrbeam.util import dict_get
 
+from octoprint_mrbeam.model.custom_material_model import CustomMaterialModel
+from octoprint_mrbeam.mrb_logger import mrb_logger
 
 # singleton
 _instance = None
 
+# TODO: SW-3719 keep these constants in a unique place
```
Comment (khaledsherkawi) [inline octoprint_mrbeam/materials.py]:
```suggestion
# TODO: SW-3719 import these from mode services
```

File: octoprint_mrbeam/__init__.py [hunk 4]
```diff
@@ -1282,16 +1274,17 @@ def custom_materials(self, data):
...
-            if "delete" in data:
-                materials(self).delete_custom_material(data["delete"])
+            if data.get("delete", []):
+                for material_key in data["delete"]:
+                    materials(self).delete_custom_material(material_key)
 
-            if "put" in data and isinstance(data["put"], dict):
-                for key, m in data["put"].iteritems():
-                    materials(self).put_custom_material(key, m)
+            if isinstance(data.get("put"), dict):
```
Comment (khaledsherkawi) [inline octoprint_mrbeam/__init__.py:1283]: This was it has a fall back that is equivalent to (if "put" in data)
```suggestion
            if isinstance(data.get("put", None), dict):
```

File: octoprint_mrbeam/__init__.py [hunk 5]
```diff
@@ -1300,6 +1293,16 @@ def custom_materials(self, data):
...
 
+    def get_laser_cutter_mode(self):
+        # TODO: SW-3719 return actual laser cutter mode
+        return "default"
+
+    def get_laser_head_model(self):
+        return self.laserhead_handler.get_current_used_lh_data()["model"]
+
+    def get_laser_head_serial(self):
+        return self.laserhead_handler.get_current_used_lh_data()["serial"]
```
Comment (khaledsherkawi) [inline octoprint_mrbeam/__init__.py]: I think these 2 methods should be get_current_laser_head... as there is more than 1 laser head model and serial (current and old)

File: octoprint_mrbeam/materials.py [hunk 6]
```diff
@@ -26,122 +29,24 @@ def __init__(self, plugin):
...
-        return res
+        return custom_materials
 
-    def reset_all_custom_materials(self):
-        self._logger.info("Resetting all custom material settings!!!!")
-        self.custom_materials = {}
-        self._save(force=True)
-
-    def _load(self, force=False):
-        if not self.custom_materials_loaded or force:
```
Comment (khaledsherkawi) [inline octoprint_mrbeam/materials.py:110]: why aren't we using this property anymore? shall we remove it?

File: octoprint_mrbeam/materials.py [hunk 7]
```diff
@@ -152,17 +57,46 @@ def _save(self, force=False):
...
-            )
-        except:
-            self._logger.exception(
-                "Exception while writing custom materials to file %s",
-                self.custom_materials_file,
+            self._logger.info("{} custom materials saved".format(len(self.custom_materials)))
+        except Exception as e:
+            self._logger.exception("Exception while saving custom materials: {}".format(e))
+
+    def get_custom_materials_for_laser_cutter_mode(self):
```
Comment (khaledsherkawi) [inline octoprint_mrbeam/materials.py:64]: Since the method doesn't have any parameters and just gets the custom material (with the current mode as it does with the current laser head) then I would not have it name "for_laser_cutter_mode"

Comment (khaledsherkawi) [reply octoprint_mrbeam/materials.py:64]: The laser head definition is still in the frontend but we will port it as some point to the backend and it should have only 1 method to get custom materials regardless of the underlaying current conditions

Comment (khaledsherkawi) [reply octoprint_mrbeam/materials.py:64]: maybe something like get_custom_materials_by_settings, more general

File: octoprint_mrbeam/materials.py [hunk 8]
```diff
@@ -1,12 +1,16 @@
 import os
```
Comment (khaledsherkawi) [inline octoprint_mrbeam/materials.py:1]: Love the refactor 😍

File: octoprint_mrbeam/model/custom_material_model.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
+
```
Comment (khaledsherkawi) [inline octoprint_mrbeam/model/custom_material_model.py:1]: Let's talk about this in a call

**Review by khaledsherkawi — CHANGES_REQUESTED** [review]

File: octoprint_mrbeam/__init__.py (hunk 4 above)
Comment (irlaec) [reply octoprint_mrbeam/__init__.py:1283]: Actually both are the same, because the default value is None: `dict.get(key, default = None)`

File: octoprint_mrbeam/__init__.py (hunk 5 above)
Comment (irlaec) [reply octoprint_mrbeam/__init__.py]: Agree, I'll change it

File: octoprint_mrbeam/materials.py (hunk 6 above)
Comment (irlaec) [reply octoprint_mrbeam/materials.py:110]: You mean self.custom_materials_loaded? Or force?
custom_materials_loaded isn't necessary because we load them in the __init__
and force was to read them from the file again, but in the current implementation it isn't necessary.
Both are removed in the current code

File: tests/test_materials.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
+from mock.mock import MagicMock
```
Comment (khaledsherkawi) [inline tests/test_materials.py:1]: Optional: Add 'get' test for an empty custom materials dict

**Review by khaledsherkawi — APPROVED** [review]

File: tests/test_materials.py (hunk 10 above)
Comment (irlaec) [reply tests/test_materials.py:1]: Done
